pyradiosky/skymodel.py

```diff
@@ -905,41 +905,52 @@ def skymodel_to_array(sky):
     catalog_table : recarray
         recarray equivalent to SkyModel data.
 
+    Notes
+    -----
+    This stores all SkyModel data in a contiguous array
+    that can be more easily handled with numpy.
+    This is used by pyuvsim for sharing catalog data via MPI.
     """
     sky.check()
     max_name_len = np.max([len(name) for name in sky.name])
-    fieldtypes = ["U" + str(max_name_len), "f8", "f8", "f8"]
-    fieldnames = ["source_id", "ra_j2000", "dec_j2000", "flux_density"]
+    fieldtypes = ["U" + str(max_name_len), "f8", "f8"]
+    fieldnames = ["source_id", "ra_j2000", "dec_j2000"]
+    stokes_names = ["I", "Q", "U", "V"]
     fieldshapes = [()] * 3
 
-    n_stokes = 4
+    n_stokes = 0
+    stokes_keep = []
+    for si, total in enumerate(np.sum(sky.stokes, axis=(1, 2))):
+        if total > 0:
+            fieldnames.append(stokes_names[si])
+            fieldshapes.append((sky.Nfreqs,))
+            fieldtypes.append("f8")
+            n_stokes += 1
+        stokes_keep.append(total > 0)
 
     if sky.freq_array is not None:
         if sky.spectral_type == "subband":
             fieldnames.append("subband_frequency")
         else:
             fieldnames.append("frequency")
         fieldtypes.append("f8")
-        fieldshapes.extend([(sky.Nfreqs, n_stokes,), (sky.Nfreqs,)])
+        fieldshapes.extend([(sky.Nfreqs,)])
```

Comment:
Does this mean you can only have Stokes I if you have a freq_array?

Reply:
It shouldn't. This line just adds the freq_array column now. The Stokes columns are added earlier and are each given a shape `(sky.Nfreqs,)`. So if you have only a single reference freq, Nfreqs=1 and the stokes column has a frequency axis of length 1.